mycroft/skills/intent_service.py

```diff
@@ -272,70 +141,75 @@ def handle_utterance(self, message):
-            padatious_matcher = PadatiousMatcher(self.padatious_service)
-
-            # List of functions to use to match the utterance with intent.
-            # These are listed in priority order.
-            match_funcs = [
-                self.converse.converse_with_skills, padatious_matcher.match_high,
-                self.adapt_service.match_intent, self.fallback.high_prio,
-                padatious_matcher.match_medium, self.fallback.medium_prio,
-                padatious_matcher.match_low, self.fallback.low_prio
-            ]
-
-            match = None
-            with stopwatch:
-                # Loop through the matching functions until a match is found.
-                for match_func in match_funcs:
-                    match = match_func(combined, lang, message)
-                    if match:
-                        break
-            if match:
-                if match.skill_id:
-                    self.converse.activate_skill(match.skill_id)
+            # iterate over intent engines (ordered by priority)
+            # matches is a list or generator with all intents found by a intent engine
+            for matches in self._match_intents(utterances, lang, message, converse=True, fallback=True):
+                # iterate over found intents, the utterance may contain more than 1 command
+                for match in matches:
                     # If the service didn't report back the skill_id it
                     # takes on the responsibility of making the skill "active"
-
-                # Launch skill if not handled by the match function
-                if match.intent_type:
-                    # keep all original message.data and update with intent
-                    # match, mycroft-core only keeps "utterances"
-                    data = dict(message.data)
-                    data.update(match.intent_data)
-                    reply = message.reply(match.intent_type, data)
-                    self.bus.emit(reply)
-
-            else:
-                # Nothing was able to handle the intent
-                # Ask politely for forgiveness for failing in this vital task
-                self.send_complete_intent_failure(message)
-            self.send_metrics(match, message.context, stopwatch)
+                    if match.skill_id:
+                        self.converse.activate_skill(match.skill_id)
+                        skill_id = match.skill_id
+                    else:
+                        # no skill_id -> no waiting for event
+                        # this usually is a fallback skill or
+                        # other handler that consumes the utterance
+                        intent_finished.set()
+
+                    # Launch skill if not handled by the match function
+                    if match.intent_type:
+                        # keep all original message.data and update with intent match
+                        # NOTE: mycroft-core only keeps "utterances"
```

Comment:
Do we want to preserve message data here? The skill should only expect the extracted intent data and original utterances for compatibility with Mycroft anyways